### PR TITLE
[WIP] Attach detach cloud volume api

### PR DESCRIPTION
**Pull request by MelsHyrule**
Fixes:
- [ ] https://github.com/ManageIQ/manageiq-ui-classic/issues/7603

dependent upon:
- [x] #1147 
- [x] #1149 

Required for upgrading [_attach.html.haml](https://github.com/ManageIQ/manageiq-ui-classic/blob/master/app/views/vm_common/_attach.html.haml#L33) and [_detach.html.haml](https://github.com/ManageIQ/manageiq-ui-classic/blob/master/app/views/vm_common/_detach.html.haml#L25)

Introduces the attach_resource and detach_resource endpoints for cloud volumes.

### CODE REVIEW

File: app/controllers/api/cloud_volumes_controller.rb [hunk 1]
```diff
@@ -47,5 +47,25 @@ def create_backup_resource(type, id, data)
         {:task_id => cloud_volume.backup_create_queue(User.current_userid, data)}
       end
     end
+
+    def attach_resource(type, id, data = {})
+      raise BadRequestError, "Must specify a vm_id" if data["vm_id"].nil?
+
+      api_resource(type, id, "Attaching Resource", :supports => :attach_volume) do |cloud_volume|
```
[the remark this answers is not in the document]
Comment (MelsHyrule) [reply app/controllers/api/cloud_volumes_controller.rb:54]: should we update `{:task_id => cloud_volume.detach_volume_queue(User.current_userid, data["vm_id"])}` to be instead ```{:task_id => cloud_volume.detach_volume_queue(User.current_userid, resource_search(data["vm_id"], :vms))}``` ?

[the remark this answers is not in the document]
Comment (agrare) [reply app/controllers/api/cloud_volumes_controller.rb:54]: `cloud_volume.attach_volume_queue` takes an `:ems_ref` not an `:id` or `Vm` object, so you should lookup the VM like @kbrock's suggestion, then pass in `vm.ems_ref`

File: app/controllers/api/cloud_volumes_controller.rb [hunk 2]
```diff
@@ -51,7 +51,26 @@ def create_backup_resource(type, id, data)
     def restore_backup_resource(type, id, data)
       api_resource(type, id, "Restoring backup for", :supports => :backup_restore) do |cloud_volume|
         {:task_id => cloud_volume.backup_restore_queue(User.current_userid, data.symbolize_keys)}
+    end
```
Comment (agrare) [inline app/controllers/api/cloud_volumes_controller.rb:54]: Missing an `end` right above this one from the conflict resolution

File: spec/requests/cloud_volumes_spec.rb [hunk 3]
```diff
@@ -192,6 +192,38 @@
...
+      expect(response).to have_http_status(:ok)
+    end
+
+    it 'detaches Cloud Volume from an instance' do
+      zone = FactoryBot.create(:zone)
+      ems = FactoryBot.create(:ems_autosde, :zone => zone)
+      hw = FactoryBot.create(:hardware)
+      vm = FactoryBot.create(:vm_vmware, :hardware => hw)
+      cloud_volume = FactoryBot.create(:cloud_volume_autosde, :ext_management_system => ems)
+      disk = FactoryBot.create(:disk, :hardware => hw, :backing => cloud_volume)
```
Comment (agrare) [inline spec/requests/cloud_volumes_spec.rb:217]: If you don't need to use `disk` later on you just need the record you could do 
```suggestion
      FactoryBot.create(:disk, :hardware => hw, :backing => cloud_volume)
```
And then rubocop won't yell at you

File: app/controllers/api/cloud_volumes_controller.rb [hunk 4]
```diff
@@ -53,5 +53,27 @@ def restore_backup_resource(type, id, data)
         {:task_id => cloud_volume.backup_restore_queue(User.current_userid, data.symbolize_keys)}
       end
     end
+
+    def attach_resource(type, id, data = {})
+      api_resource(type, id, "Attaching Resource to", :supports => :attach_volume) do |cloud_volume|
+        raise BadRequestError, "Must specify a vm_id" if data["vm_id"].blank?
+
+        vm = resource_search(data["vm_id"], :vms)
+        {:task_id => cloud_volume.attach_volume_queue(User.current_userid, vm.ems_ref)}
```
Comment (agrare) [inline app/controllers/api/cloud_volumes_controller.rb:62]: @MelsHyrule we'll also need an optional `device` parameter here:
```suggestion
        {:task_id => cloud_volume.attach_volume_queue(User.current_userid, vm.ems_ref, data["device"].presence)}
```
should do the trick

File: spec/requests/cloud_volumes_spec.rb [hunk 5]
```diff
@@ -192,6 +192,38 @@
...
     end
+
+    it 'attaches Cloud Volume to an instance' do
+      zone = FactoryBot.create(:zone)
+      ems = FactoryBot.create(:ems_autosde, :zone => zone)
+      vm = FactoryBot.create(:vm_vmware)
+      cloud_volume = FactoryBot.create(:cloud_volume_autosde, :ext_management_system => ems)
+
+      api_basic_authorize(action_identifier(:cloud_volumes, :attach, :resource_actions, :post))
+      stub_supports(cloud_volume.class, :attach)
```
Comment (agrare) [inline spec/requests/cloud_volumes_spec.rb:203]:
```suggestion
      stub_supports(cloud_volume.class, :attach_volume)
```

File: spec/requests/cloud_volumes_spec.rb [hunk 6]
```diff
@@ -192,6 +192,38 @@
...
+    it 'detaches Cloud Volume from an instance' do
+      zone = FactoryBot.create(:zone)
+      ems = FactoryBot.create(:ems_autosde, :zone => zone)
+      hw = FactoryBot.create(:hardware)
+      vm = FactoryBot.create(:vm_vmware, :hardware => hw)
+      cloud_volume = FactoryBot.create(:cloud_volume_autosde, :ext_management_system => ems)
+      FactoryBot.create(:disk, :hardware => hw, :backing => cloud_volume)
+
+      api_basic_authorize(action_identifier(:cloud_volumes, :detach, :resource_actions, :post))
+      stub_supports(cloud_volume.class, :detach)
```
Comment (agrare) [inline spec/requests/cloud_volumes_spec.rb:220]:
```suggestion
      stub_supports(cloud_volume.class, :detach_volume)
```

File: app/controllers/api/cloud_volumes_controller.rb [hunk 7]
```diff
@@ -34,14 +34,48 @@ def delete_resource_main_action(_type, cloud_volume, _data)
...
         super
       end
     end
 
+    def render_attach_resource_options_mels(ems_id)
+      p "-------------------------- mels attach options "
+      type = @req.collection.to_sym
+      base_klass = collection_class(type)
+
+      ems = resource_search(ems_id, :providers)
```
Comment (kbrock) [inline app/controllers/api/cloud_volumes_controller.rb:56]: ems may be nil

File: app/controllers/api/base_controller/renderer.rb [hunk 8]
```diff
@@ -549,12 +549,20 @@ def render_options(resource, data = {})
...
       end
 
-      def render_update_resource_options(id)
+      def render_update_resource_options(id, action)
         type = @req.collection.to_sym
         resource = resource_search(id, type)
         raise BadRequestError, resource.unsupported_reason(:update) unless resource.supports?(:update)
 
-        render_options(type, :form_schema => resource.params_for_update)
+        if action == "attach"
```
Comment (agrare) [inline app/controllers/api/base_controller/renderer.rb:557]: I don't think overriding the `render_update_resource_options` method is the way to go here

For example won't we not get here if we are looking for the "attach" action but the resource doesn't support "update" a totally different action? https://github.com/ManageIQ/manageiq-api/pull/1141/files#diff-48fb0a597ebc41e5c8c4fde22100d5577a08a07b84761152f69eaa2f580c4c37R555

Plus "update" is in the name, then we also have it handle other options

File: app/controllers/api/cloud_volumes_controller.rb [hunk 9]
```diff
@@ -34,7 +34,8 @@ def delete_resource_main_action(_type, cloud_volume, _data)
 
     def options
       if (id = params["id"])
-        render_update_resource_options(id)
+        action = params["option_action"] || "update"
+        render_update_resource_options(id, action)
```
Comment (agrare) [inline app/controllers/api/cloud_volumes_controller.rb:38]: Could we `send("render_#{action}_resource_options(id)")` instead?  Then we would localize the methods to the controller that has those actions

**Comment by agrare** [comment]
@MelsHyrule I was good with tackling the `params_for_attach` in a separate PR (https://github.com/ManageIQ/manageiq-ui-classic/pull/8205#discussion_r841804705) I think what you had prior to trying to add the OPTIONS call was good if you want to move forward with that and I think @kbrock is tackling a general way of dealing with action options in his other PR

**Comment by agrare** [comment]
@MelsHyrule merged https://github.com/ManageIQ/manageiq-api/pull/1147 if you want to build on that

**Comment by MelsHyrule** [comment]
With changes from https://github.com/ManageIQ/manageiq-api/pull/1147 this PR is no longer needed, closing now.

**Pull request closed without merging**